### PR TITLE
Feature/api/device-token

### LINKED ISSUE #798: Criar rota para armazenar tokens dos dispositivos para notificações push
## Feature Description

Criar rota para armazenar tokens dos dispositivos para notificações push

---------------

## Acceptance criteria

* 

## Implementation Brief

*

### PR DESCRIPTION
**Pull request by alexandreadames**
## Summary

Resolves #798 

## Relevant technical choices

https://www.loom.com/share/23598168473a4f168137f31cc6cae1a0

### Rotas

| Description | Method | Route | Body/Query | Auth |
| --- | --- | --- | --- | --- |
Lists All User Device Tokens|GET|device-tokens | N/A| YES | 
Gets a single User Device Token|GET|device-tokens/:uuid | N/A| YES | 
Creates a new User Device Token|POST|device-tokens |device_uuid(String)<br>device_token(string)| YES | 
Updates a User Device Token|PUT|device-tokens/:uuid |device_token(string)| YES | 
Deletes a Device Token|DELETE|device-tokens/:uuid |N/A| YES | 
Deletes user Device Tokens|DELETE|device-tokens |N/A| YES | 

## QA Steps

Run `test/functional/device-token.spec.js`

## Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [x] My code passes the linting.
- [x] My code has proper inline documentation.
- [x] I have manually tested this PR.

### CODE REVIEW

File: packages/api/app/Controllers/Http/DeviceTokenController.js [hunk 1]
```diff
@@ -0,0 +1,95 @@
+const DeviceToken = use('App/Models/DeviceToken');
+const { errorPayload, errors } = require('../../Utils');
+
+class DeviceTokenController {
+	async index({ auth }) {
+		const deviceOwner = await auth.getUser();
+		const deviceTokens = await DeviceToken.query()
+			.where({ user_id: deviceOwner.id })
+			.fetch();
+		return deviceTokens;
```
Comment (mateus4k) [inline packages/api/app/Controllers/Http/DeviceTokenController.js:10]:
```suggestion
		return User.deviceTokens().fetch();
```

File: packages/api/app/Controllers/Http/DeviceTokenController.js [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+		if (!deviceToken) {
+			return response
+				.status(403)
+				.send(
+					errorPayload(
+						errors.UNAUTHORIZED_ACCESS,
+						request.antl('error.permission.unauthorizedAccess'),
+					),
+				);
+		}
```
Comment (mateus4k) [inline packages/api/app/Controllers/Http/DeviceTokenController.js:40]: Isso pode ser um `exists` no validator 🤔

File: packages/api/app/Controllers/Http/DeviceTokenController.js [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+	}
+
+	async show({ auth, params }) {
+		const { uuid } = params;
+		const deviceOwner = await auth.getUser();
+		const deviceToken = await DeviceToken.query()
+			.where({ device_uuid: uuid })
+			.where({ user_id: deviceOwner.id })
+			.first();
+		return deviceToken;
```
Comment (mateus4k) [inline packages/api/app/Controllers/Http/DeviceTokenController.js:53]:
```suggestion
		return User.deviceTokens()
			.query()
			.where({ device_uuid: params.uuid })
			.first();
```

File: packages/api/database/migrations/1612300913780_device_token_schema.js [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+class DeviceTokenSchema extends Schema {
+	up() {
+		this.create('device_tokens', (table) => {
+			table.increments();
+			table
+				.integer('user_id')
+				.unsigned()
+				.references('id')
+				.inTable('users');
```
Comment (mateus4k) [inline packages/api/database/migrations/1612300913780_device_token_schema.js:12]: É interessante apagarmos os tokens ao excluirmos o usuário
```suggestion
				.inTable('users')
				.onUpdate('cascade')
				.onDelete('cascade');
```

**Review by mateus4k — CHANGES_REQUESTED** [review]

File: packages/api/app/Controllers/Http/DeviceTokenController.js (hunk 2 above)
Comment (alexandreadames) [reply packages/api/app/Controllers/Http/DeviceTokenController.js:40]: O ideal aqui seria criar uma permissão nova, mas eu quis simplificar. A lógica é não permitir que outro usuário atualize o token criado por outro.

**Review by mateus4k — DISMISSED** [review]

**Comment by mateus4k** [comment]
@alexandreadames please solve conflicts

**Comment by alexandreadames** [comment]
> @alexandreadames please solve conflicts

Solved!

**Review by mateus4k — APPROVED** [review]

**Review by mtullyoc — APPROVED** [review]
Great!

**Pull request merged**